File: runtimes/java/src/hime/redist/parsers/BaseLRParser.java
/**********************************************************************
 * Copyright (c) 2014 Laurent Wouters and others
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General
 * Public License along with this program.
 * If not, see <http://www.gnu.org/licenses/>.
 *
 * Contributors:
 *     Laurent Wouters - lwouters@xowl.org
 **********************************************************************/
package hime.redist.parsers;

import hime.redist.ParseError;
import hime.redist.SemanticAction;
import hime.redist.Symbol;
import hime.redist.lexer.ILexer;
import hime.redist.lexer.LexicalErrorHandler;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * Represents a base LR parser
 */
public abstract class BaseLRParser implements IParser {
    /**
     * Maximal size of the stack
     */
    protected static final int maxStackSize = 128;
    /**
     * Maximum number of errors
     */
    protected static final int maxErrorCount = 100;

    /**
     * Determines whether the parser will try to recover from errors
     */
    protected boolean recover = true;
    /**
     * Parser's variables
     */
    protected List<Symbol> parserVariables;
    /**
     * Parser's virtuals
     */
    protected List<Symbol> parserVirtuals;
    /**
     * Parser's actions
     */
    protected SemanticAction[] parserActions;
    /**
     * List of the encountered syntaxic errors
     */
    protected List<ParseError> allErrors;
    /**
     * Lexer associated to this parser
     */
    protected ILexer lexer;


    /**
     * Gets the variable symbols used by this parser
     *
     * @return The variable symbols used by this parser
     */
    public List<Symbol> getVariables() {
        return parserVariables;
    }

    /**
     * Gets the virtual symbols used by this parser
     *
     * @return The virtual symbols used by this parser
     */
    public List<Symbol> getVirtuals() {
        return parserVirtuals;
    }

    /**
     * Gets whether the parser should try to recover from errors
     *
     * @return Whether the parser should try to recover from errors
     */
    public boolean getRecover() {
        return recover;
    }

    /**
     * Sets whether the parser should try to recover from errors
     *
     * @param recover Whether the parser should try to recover from errors
     */
    public void setRecover(boolean recover) {
        this.recover = recover;
    }

    /**
     * Initializes a new instance of the LRkParser class with the given lexer
     *
     * @param variables The parser's variable
     * @param virtuals  The parser's virtuals
     * @param actions   The parser's actions
     * @param lexer     The input lexer
     */
    protected BaseLRParser(Symbol[] variables, Symbol[] virtuals, SemanticAction[] actions, ILexer lexer) {
        this.parserVariables = Collections.unmodifiableList(Arrays.asList(variables));
        this.parserVirtuals = Collections.unmodifiableList(Arrays.asList(virtuals));
        this.parserActions = actions;
        this.recover = true;
        this.allErrors = new ArrayList<ParseError>();
        this.lexer = lexer;
        this.lexer.setErrorHandler(new LexicalErrorHandler() {
            @Override
            public void handle(ParseError error) {
                allErrors.add(error);
            }
        });
    }
}

File: runtimes/net/Sources/Parsers/LR/BaseLRParser.cs
/**********************************************************************
* Copyright (c) 2013 Laurent Wouters and others
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License as
* published by the Free Software Foundation, either version 3
* of the License, or (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General
* Public License along with this program.
* If not, see <http://www.gnu.org/licenses/>.
*
* Contributors:
*     Laurent Wouters - lwouters@xowl.org
**********************************************************************/
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;

namespace Hime.Redist.Parsers
{
	/// <summary>
	/// Represents a base LR parser
	/// </summary>
	public abstract class BaseLRParser : IParser
	{
		/// <summary>
		/// Maximal size of the stack
		/// </summary>
		protected const int maxStackSize = 128;
		/// <summary>
		/// Maximum number of errors
		/// </summary>
		protected const int maxErrorCount = 100;

		/// <summary>
		/// Determines whether the parser will try to recover from errors
		/// </summary>
		protected bool recover = true;
		/// <summary>
		/// Parser's variables
		/// </summary>
		protected IList<Symbol> parserVariables;
		/// <summary>
		/// Parser's virtuals
		/// </summary>
		protected IList<Symbol> parserVirtuals;
		/// <summary>
		/// Parser's actions
		/// </summary>
		protected SemanticAction[] parserActions;
		/// <summary>
		/// List of the encountered syntaxic errors
		/// </summary>
		protected List<ParseError> allErrors;
		/// <summary>
		/// Lexer associated to this parser
		/// </summary>
		protected Lexer.ILexer lexer;


		/// <summary>
		/// Gets the variable symbols used by this parser
		/// </summary>
		public IList<Symbol> Variables { get { return parserVariables; } }
		/// <summary>
		/// Gets the virtual symbols used by this parser
		/// </summary>
		public IList<Symbol> Virtuals { get { return parserVirtuals; } }
		/// <summary>
		/// Gets or sets whether the paser should try to recover from errors
		/// </summary>
		public bool RecoverErrors
		{
			get { return recover; }
			set { recover = value; }
		}

		/// <summary>
		/// Initializes a new instance of the LRkParser class with the given lexer
		/// </summary>
		/// <param name="variables">The parser's variables</param>
		/// <param name="virtuals">The parser's virtuals</param>
		/// <param name="actions">The parser's actions</param>
		/// <param name="lexer">The input lexer</param>
		protected BaseLRParser(Symbol[] variables, Symbol[] virtuals, SemanticAction[] actions, Lexer.ILexer lexer)
		{
			this.parserVariables = new ReadOnlyCollection<Symbol>(new List<Symbol>(variables));
			this.parserVirtuals = new ReadOnlyCollection<Symbol>(new List<Symbol>(virtuals));
			this.parserActions = actions;
			this.recover = true;
			this.allErrors = new List<ParseError>();
			this.lexer = lexer;
			this.lexer.OnError += OnLexicalError;
		}

		/// <summary>
		/// Adds the given lexical error emanating from the lexer to the list of errors
		/// </summary>
		/// <param name="error">Lexical error</param>
		protected void OnLexicalError(ParseError error)
		{
			allErrors.Add(error);
		}

		/// <summary>
		/// Parses the input and returns the result
		/// </summary>
		/// <returns>A ParseResult object containing the data about the result</returns>
		public abstract ParseResult Parse();
	}
}
